sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -190,9 +191,9 @@ Connection getConnection() throws Exception {
         basicDataSource.setPassword(getPassword());
         dataSource = basicDataSource;
       }
-      return (getUsername() == null)
-          ? dataSource.getConnection()
-          : dataSource.getConnection(getUsername(), getPassword());
+      return (getDataSource() != null && getUsername() != null)
```

Comment:
OK, on further look, this looks incorrect: if getDataSource() != null, then getUsername() == null and getPassword() == null, by construction (there is no create() method that fills in datasource AND username/password). So the condition is always false and you're always returning dataSource.getConnection().

If I understood this right and it's actually a bug, can you add a test for this situation?

And, I still believe that the code will be *much* easier to read if you put 1 return statement in each if/else branch above, instead of duplicating the check of whether or not a datasource was supplied.